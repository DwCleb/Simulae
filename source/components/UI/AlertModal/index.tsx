import React from 'react';

import { Container, Content, Title, Message, Button, ButtonText } from './styles';

interface AlertModalProps {
  onClose: (status) => void;
  modalVisible: boolean;
  title: string;
  message: string;
}

function AlertModal({ onClose, modalVisible, title, message }: AlertModalProps): JSX.Element {
  return (
    <Container isVisible={modalVisible} swipeDirection="down" onBackdropPress={(): void => onClose(!modalVisible)}>
      <Content>
        <Title>{title}</Title>
        <Message>{message}</Message>
        <Button onPress={(): void => onClose(!modalVisible)}>
          <ButtonText>Okay</ButtonText>
        </Button>
      </Content>
    </Container>
  );
}

export default AlertModal;
