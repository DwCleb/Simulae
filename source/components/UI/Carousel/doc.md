## <a name="section-carousel"> Carousel </a>

### Description

A carousel component to render diferent items in template.

### Props

| Name               | Type       | Required | Default | Description                                                  |
| ------------------ | ---------- | -------- | ------- | ------------------------------------------------------------ |
| data               | `Object[]` | `true`   | -       | Array with data that will be render inside item by template. |
| templateItem       | `String`   | `false`  | explore | Set item template that will be render                        |
| templatePagination | `String`   | `false`  | explore | Set pagination template that will be render                  |

### Usage example

```js
  ...
  const DATA = [
    {
      title: 'Welcome to Jawwy',
      image: Logo,
      description: 'Join us! Personalise, track and manage your prepaid plan in real time.',
    },
    {
      title: 'Your plan, your way',
      image: Step2,
      description: 'We give you Unlimited Voice and SMS.\nJust select the data, add-ons and validity.',
    }
  ];

  return <Carousel data={DATA} />
```
