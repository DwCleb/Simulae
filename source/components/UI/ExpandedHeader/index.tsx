import React from 'react';

import { BackButton } from '~/components/Navigation';
import { Container, BackButtonContainer, TextContainer, Title, Subtitle } from './styles';

interface ExpandedHeaderProps {
  title: string;
  subtitle?: string | boolean;
  activeBack?: boolean;
  actionBack?: () => void;
}
export default function ExpandedHeader({ title, subtitle, activeBack, actionBack }: ExpandedHeaderProps): JSX.Element {
  return (
    <Container>
      {activeBack && (
        <BackButtonContainer>
          <BackButton action={actionBack} />
        </BackButtonContainer>
      )}
      <TextContainer>
        <Title>{title}</Title>
        {subtitle && <Subtitle>{subtitle}</Subtitle>}
      </TextContainer>
    </Container>
  );
}
