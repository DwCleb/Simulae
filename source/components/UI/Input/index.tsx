import React, { forwardRef, useState, useEffect } from 'react';
import { Animated, View } from 'react-native';
import PropTypes from 'prop-types';
import { MaskedInput, TextInput } from './styles';

import { Colors } from '~/theme';

function Input(
  {
    dark,
    name,
    distance,
    onChange,
    label,
    type,
    value,
    weight,
    placeholder,
    accessibilityLabel,
    keyboardType,
    ...rest
  },
  ref
): JSX.Element {
  accessibilityLabel = `Campo de texto ${label || placeholder || name}`;

  const [isFocused, setIsFocused] = useState(false);

  const animatedIsFocused = new Animated.Value(value || isFocused ? 1 : 0);

  useEffect(() => {
    Animated.timing(animatedIsFocused, {
      toValue: (isFocused || value !== undefined) && (isFocused || value !== '') ? 1 : 0,
      duration: 200,
      useNativeDriver: false,
    }).start();
  }, [animatedIsFocused, isFocused, name, value]);

  const labelStyle = {
    position: 'absolute',
    left: 2,
    top: animatedIsFocused.interpolate({
      inputRange: [0, 1],
      outputRange: [distance ? distance[1] + 25 : 25, 8],
    }),
    fontSize: animatedIsFocused.interpolate({
      inputRange: [0, 1],
      outputRange: [14, 12],
    }),
    color: animatedIsFocused.interpolate({
      inputRange: [0, 1],
      outputRange: dark
        ? [Colors.theme.primary.purple, Colors.theme.primary.purple]
        : [Colors.theme.monochromatic.lightGrey, Colors.theme.monochromatic.altGrey],
    }),
  };

  function handleFocus(): void {
    Animated.timing(animatedIsFocused, {
      toValue: 1,
      duration: 200,
      useNativeDriver: false,
    }).start();
    setTimeout(() => {
      setIsFocused(true);
    }, 200);
  }

  function handleBlur(): void {
    Animated.timing(animatedIsFocused, {
      toValue: value ? 1 : 0,
      duration: 200,
      useNativeDriver: false,
    }).start();
    setTimeout(() => {
      setIsFocused(false);
    }, 200);
  }

  return (
    <View style={{ marginTop: 20 }}>
      <Animated.Text style={labelStyle}>{label}</Animated.Text>
      {type ? (
        <MaskedInput
          dark={dark}
          value={value}
          onChangeText={(text: string): void => onChange({ value: text, name })}
          placeholder={placeholder}
          type={type}
          {...rest}
          ref={ref}
          onFocus={(): void => handleFocus()}
          onBlur={(): void => handleBlur()}
          keyboardType={keyboardType}
          accessibilityLabel={accessibilityLabel}
        />
      ) : (
        <TextInput
          dark={dark}
          value={value}
          onChangeText={(text: string): void => onChange({ value: text, name })}
          placeholder={placeholder}
          {...rest}
          ref={ref}
          onFocus={(): void => handleFocus()}
          onBlur={(): void => handleBlur()}
          keyboardType={keyboardType}
          accessibilityLabel={accessibilityLabel}
        />
      )}
      {/* <View style={{ marginTop: 2, marginBottom: 10 }}>
        <Typography type={TypographyType.L1}></Typography>
      </View> */}
    </View>
  );
}

Input.defaultProps = {
  type: '',
  weight: false,
  keyboardType: '',
};

Input.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  type: PropTypes.string,
  weight: PropTypes.number,
  placeholder: PropTypes.string.isRequired,
  accessibilityLabel: PropTypes.string.isRequired,
  keyboardType: PropTypes.string,
  distance: PropTypes.number.isRequired,
};

export default forwardRef(Input);
