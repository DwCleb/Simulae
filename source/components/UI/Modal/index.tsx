import React from 'react';

import { Container, Content, ModalRod } from './styles';

function Modal({ setModalVisible, modalVisible, children }): JSX.Element {
  return (
    <Container
      isVisible={modalVisible}
      avoidKeyboard={true}
      swipeDirection="down"
      swipeThreshold={60}
      onBackdropPress={() => setModalVisible(!modalVisible)}
      onSwipeComplete={() => setModalVisible(!modalVisible)}
    >
      <Content>
        <ModalRod />
        {children}
      </Content>
    </Container>
  );
}

export default Modal;
