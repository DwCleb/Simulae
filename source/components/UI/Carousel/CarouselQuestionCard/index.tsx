import React from 'react';
import { Platform } from 'react-native';

import { Typography } from '~/components';
import { TypographyType } from '~/components/UI/Typography';

import { Container, TitleContainer, IllustrationContainer, DescriptionContainer, CarouselImage } from './styles';

export interface CarouselItemProps {
  item: { name: string; checkIn: string; image: JSX.Element; overlap: boolean; cardIndex: number };
}

const overlapTitleStyle = { zIndex: 999, marginTop: Platform.select({ android: -30, ios: -60 }) };
const overlapImageWidth = Platform.select<string | number>({ android: '135%', ios: 430 });
const overlapImageHeight = Platform.select<string | number>({ android: '135%', ios: 400 });

export default function CarouselQuestionCard({ item, cardIndex }: CarouselItemProps): JSX.Element {
  const { name, checkIn, image, overlap } = item;

  return (
    <Container cardIndex={cardIndex}>
      <TitleContainer>
        <Typography type={TypographyType.H2} distance={[20, 0, 0, 0]}>
          {name}
        </Typography>
      </TitleContainer>
      <IllustrationContainer>
        <CarouselImage
          source={image}
          width={overlap ? overlapImageWidth : '100%'}
          height={overlap ? overlapImageHeight : '100%'}
          style={overlap ? overlapTitleStyle : {}}
        />
      </IllustrationContainer>

      <DescriptionContainer>
        <Typography type={TypographyType.L2} distance={[0, 0, 0, 0]}>
          {checkIn}
        </Typography>
      </DescriptionContainer>
    </Container>
  );
}
