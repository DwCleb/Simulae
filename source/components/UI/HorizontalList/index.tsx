import React from 'react';
import { FlatList } from 'react-native';

import { Typography } from '~/components';
import { TypographyType } from '~/components/UI/Typography';

export default function HorizontalList({ ListItem, headerTitle, data }): JSX.Element {
  return (
    <React.Fragment>
      <Typography type={TypographyType.H3} distance={[0, 0, 0, 20]} align="justify">
        {headerTitle}
      </Typography>
      <FlatList
        data={data}
        horizontal
        showsHorizontalScrollIndicator={false}
        renderItem={({ item, index }): JSX.Element => {
          console.log(index);
          const firstItemIndex = index === 0;
          return <ListItem item={{ ...item }} isFirstItemIndex={firstItemIndex} onPress={(): void => {}} />;
        }}
        keyExtractor={(item): string => item.id}
      />
    </React.Fragment>
  );
}
