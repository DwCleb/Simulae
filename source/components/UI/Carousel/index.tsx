import React, { useRef, useState, Fragment } from 'react';
import Carousel from 'react-native-snap-carousel';

import { Metrics } from '~/theme';

export interface CarouselItemProps {
  title: string;
  description: string;
  image: JSX.Element;
  overlap: boolean;
}

export interface CarouselProps {
  data: object[];
  CarouselItem: React.FC<any>;
  Pagination?: React.FC<{ length: number; activeDotIndex: number }>;
  moveOnPress?: boolean;
  loop?: boolean;
  selectedTab?: number;
  handleChange?: (index: number) => void;
}

export default function CustomCarousel({
  data,
  CarouselItem,
  Pagination,
  moveOnPress,
  loop,
  selectedTab,
  handleChange,
}: CarouselProps): JSX.Element {
  const carouselRef = useRef<any>(null);

  React.useEffect(() => {
    if (carouselRef && carouselRef.current) {
      carouselRef.current.snapToItem(selectedTab);
    }
  }, [selectedTab]);

  const { width } = Metrics.screen;

  const [sliderActiveSlide, setSliderActiveSlide] = useState(0);

  function snapToNextSlide(): void {
    if (carouselRef && carouselRef.current) {
      carouselRef.current.snapToNext();
    }
  }

  const handleSlideChange = (index: number): void => {
    if (handleChange) {
      handleChange(index);
    }
    setSliderActiveSlide(index);
  };

  return (
    <Fragment>
      {Pagination && <Pagination length={data.length} activeDotIndex={sliderActiveSlide} />}
      <Carousel
        ref={carouselRef}
        sliderWidth={width}
        sliderHeight={500}
        itemWidth={width - 40}
        data={data}
        renderItem={({ item, index }): JSX.Element => (
          <CarouselItem item={{ ...item }} cardIndex={index} onPress={moveOnPress && snapToNextSlide} />
        )}
        onSnapToItem={handleSlideChange}
        initialNumToRender={data.length}
        firstItem={0}
        loop={loop}
        loopClonesPerSide={1}
      />
      {Pagination && <Pagination length={data.length} activeDotIndex={sliderActiveSlide} />}
    </Fragment>
  );
}
