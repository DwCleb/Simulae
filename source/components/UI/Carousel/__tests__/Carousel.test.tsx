import React from 'react';
import { Text } from 'react-native';
import { create, ReactTestRenderer, act } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';

import { Carousel } from '~/components';

import { CarouselProps } from '~/components/UI/Carousel';

import darkTheme from '~/theme/ThemeContext/dark';

let wrapper: ReactTestRenderer;
let root: ReactTestRenderer['root'];

const FakeCarouselItem = ({ item }: { item: { label: string } }): JSX.Element => <Text>{item.label}</Text>;
const FakePagination = (): JSX.Element => <Text>Pagination</Text>;

const handleChangeEvent = jest.fn();

const props: CarouselProps = { data: [{ label: 'mock 1' }, { label: 'mock 2' }], CarouselItem: FakeCarouselItem };

const theme = darkTheme.theme;

const componentToTest = (componentProps: CarouselProps): JSX.Element => (
  <ThemeProvider theme={theme}>
    <Carousel {...componentProps} />
  </ThemeProvider>
);

describe('<Carousel />', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  beforeEach(() => {
    wrapper = create(componentToTest(props));
    root = wrapper.root;
  });

  it('Should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render all the necessary components - mandatory props', () => {
    expect(root.findAllByType(Carousel)).toHaveLength(1);

    expect(root.findAllByType(FakeCarouselItem)).toHaveLength(props.data.length);
  });

  it('Should render with pagination prop', () => {
    const newProps = { ...props, Pagination: FakePagination };

    wrapper = create(componentToTest(newProps));
    root = wrapper.root;

    expect(root.findAllByType(Carousel)).toHaveLength(1);
    expect(root.findAllByType(Text)).toHaveLength(props.data.length + 1);
    expect(root.findAllByType(Text)[2].props.children).toEqual('Pagination');
  });

  it('Should trigger handleChange function if carousel onSnapToItem happens', () => {
    const newProps = { ...props, handleChange: handleChangeEvent };

    wrapper = create(componentToTest(newProps));
    root = wrapper.root;

    //TODO: Review this type, it works but its not intended to be like this
    //@ts-ignore
    const carouselElement = root.findAllByType('RCTScrollView')[0];

    act(() => {
      carouselElement.props.onSnapToItem(1);
    });

    expect(handleChangeEvent).toHaveBeenCalledWith(1);
  });
});
