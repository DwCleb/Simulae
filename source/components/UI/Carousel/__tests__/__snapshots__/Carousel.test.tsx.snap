// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> Should match the snapshot 1`] = `
<RCTScrollView
  activeAnimationOptions={null}
  activeAnimationType="timing"
  activeSlideAlignment="center"
  activeSlideOffset={20}
  apparitionDelay={0}
  automaticallyAdjustContentInsets={false}
  autoplay={false}
  autoplayDelay={1000}
  autoplayInterval={3000}
  callbackOffsetMargin={5}
  containerCustomStyle={Object {}}
  contentContainerCustomStyle={Object {}}
  contentContainerStyle={
    Array [
      Object {
        "paddingLeft": 0,
        "paddingRight": 0,
      },
      Object {},
    ]
  }
  data={
    Array [
      Object {
        "label": "mock 1",
      },
      Object {
        "label": "mock 2",
      },
    ]
  }
  decelerationRate="fast"
  directionalLockEnabled={true}
  disableVirtualization={false}
  enableMomentum={false}
  enableSnap={true}
  firstItem={0}
  forwardedRef={[Function]}
  getItem={[Function]}
  getItemCount={[Function]}
  hasParallaxImages={false}
  horizontal={true}
  inactiveSlideOpacity={0.7}
  inactiveSlideScale={0.9}
  inactiveSlideShift={0}
  initialNumToRender={2}
  invertStickyHeaders={false}
  inverted={false}
  itemWidth={750}
  keyExtractor={[Function]}
  layout="default"
  lockScrollTimeoutDuration={1000}
  lockScrollWhileSnapping={false}
  loop={false}
  loopClonesPerSide={1}
  maxToRenderPerBatch={13}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onResponderRelease={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onSnapToItem={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  overScrollMode="never"
  pinchGestureEnabled={false}
  removeClippedSubviews={true}
  renderItem={[Function]}
  scrollEnabled={true}
  scrollEventThrottle={1}
  scrollsToTop={false}
  shouldOptimizeUpdates={true}
  showsHorizontalScrollIndicator={false}
  showsVerticalScrollIndicator={false}
  slideStyle={Object {}}
  sliderHeight={750}
  sliderWidth={750}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "flexDirection": "row",
      "opacity": 0,
      "width": 750,
    }
  }
  swipeThreshold={20}
  updateCellsBatchingPeriod={50}
  useScrollView={false}
  vertical={false}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={13}
>
  <View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "opacity": 1,
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": 750,
          }
        }
      >
        <Text>
          mock 1
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "opacity": 0.7,
            "transform": Array [
              Object {
                "scale": 0.9,
              },
            ],
            "width": 750,
          }
        }
      >
        <Text>
          mock 2
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
